smu/geometry/topology_from_geom.py

```diff
@@ -0,0 +1,174 @@
+    if bond_topology.atoms[a1] != dataset_pb2.BondTopology.AtomType.ATOM_H:
+      continue
+
+    shortest_distance = 1.0e+30
+    closest_heavy_atom = -1
+    for a2 in range(0, natoms):
+      if bond_topology.atoms[a2] == dataset_pb2.BondTopology.AtomType.ATOM_H:
+        continue
+
+      if distances[a1, a2] >= THRESHOLD:
+        continue
+
+      if distances[a1, a2] < shortest_distance:
+        shortest_distance = distances[a1, a2]
+        closest_heavy_atom = a2
+
+    if closest_heavy_atom < 0:
+      return None
+
+    bond = dataset_pb2.BondTopology.Bond(atom_a=a1,
+                                         atom_b=closest_heavy_atom,
+                                         bond_type=dataset_pb2.BondTopology.BondType.BOND_SINGLE)
+    result.bonds.append(bond)
+
+  return result
+
+
+def indices_of_heavy_atoms(bond_topology: dataset_pb2.BondTopology) -> List[int]:
+  """Return the indices of the heavy atoms in `bond_topology`.
+    Args:
+    Returns:
+  """
+  return [
+      i for i, t in enumerate(bond_topology.atoms) if t != dataset_pb2.BondTopology.AtomType.ATOM_H
+  ]
+
+
+def bond_topologies_from_geom(
+    bond_lengths: bond_length_distribution.AllAtomPairLengthDistributions,
+    bond_topology: dataset_pb2.BondTopology, geometry: dataset_pb2.Geometry,
+    matching_parameters: smu_molecule.MatchingParameters) -> dataset_pb2.TopologyMatches:
+  """Return all BondTopology's that are plausible.
+
+    Given a molecule described by `bond_topology` and `geometry`, return all possible
+    BondTopology that are consistent with that.
+    Note that `bond_topology` will be put in a canonical form.
+    Args:
+      bond_length_distribution:
+      bond_topology:
+      geometry:
+    Returns:
+      TopologyMatches
+  """
+  result = dataset_pb2.TopologyMatches()    # To be returned.
+  if len(bond_topology.atoms) == 1:
+    return result    # empty.
+
+  # Will be used when comparing perceived BondTopology's.
+  serialized_starting_form = bond_topology.SerializeToString()
+
+  utilities.canonical_bond_topology(bond_topology)
+  distances = utilities.distances(geometry)
+
+  # First join each Hydrogen to its nearest heavy atom, thereby
+  # creating a starting BondTopology from which all others can grow
+  starting_bond_topology = hydrogen_to_nearest_atom(bond_topology, distances)
+  if starting_bond_topology is None:
+    return result
+
+# heavy_atoms = [a for a in bond_topology.atoms if a != dataset_pb2.BondTopology.AtomType.ATOM_H]
+# heavy_atom_indices = indices_of_heavy_atoms(bond_topology)
+  heavy_atom_indices = [
+      i for i, t in enumerate(bond_topology.atoms) if t != dataset_pb2.BondTopology.AtomType.ATOM_H
+  ]
+
+  # For each atom pair, a list of possible bond types.
+  # Key is a tuple of the two atom numbers, value is an np.array
+  # with the score for each bond type.
+
+  bonds_to_scores: Dict[Tuple[int, int], np.array] = {}
+  for c in itertools.combinations(heavy_atom_indices, 2):    # All pairs.
```

Comment:
for i, j in itertools.combinations(...)
is simpler.

Reply:
Much nicer, yes.

When I wrote this in Julia, that's how I did it...